### PR TITLE
[Abandoned] Fix number of lines reported by function_body_length and type_body_length

### PR DESCRIPTION
**Pull request by ornithocoder**
This commit implements a fix for the issue described in #1750. The following rules are returning the incorrect number of lines:

* `function_body_length`
* `type_body_length`

This PR contains several commits to make the code review easier.

### CODE REVIEW

**Comment by marcelofabri** [comment]
Reminder to my future self: run `oss-check` locally to see if this impacts the performance.

**Comment by ornithocoder** [comment]
@marcelofabri I got this:

```
personal-fork-swiftlint $ ./script/oss-check --branch blank_line --iteration 10
Preparing osscheck/builds (identifier builds)
Previous HEAD position was 81fb8a64... Add test for case which was reporting incorrect number of lines
HEAD is now at ef67adca... rename RulesDocsOptions to GenerateDocsOptions
Message: Linting Aerial with this PR took 0.3s vs 0.31s on master (3% faster)
Message: Linting Alamofire with this PR took 2.25s vs 2.34s on master (3% faster)
Message: Linting Firefox with this PR took 9.57s vs 10.55s on master (9% faster)
Message: Linting Kickstarter with this PR took 13.84s vs 13.91s on master (0% faster)
Message: Linting Moya with this PR took 0.97s vs 0.63s on master (53% slower)
Message: Linting Nimble with this PR took 1.2s vs 1.25s on master (4% faster)
Message: Linting Quick with this PR took 0.4s vs 0.42s on master (4% faster)
Message: Linting Realm with this PR took 2.08s vs 2.02s on master (2% slower)
Message: Linting SourceKitten with this PR took 0.75s vs 0.72s on master (4% slower)
Message: Linting Sourcery with this PR took 3.59s vs 3.75s on master (4% faster)
Message: Linting Swift with this PR took 11.08s vs 11.13s on master (0% faster)
Message: Linting WordPress with this PR took 9.42s vs 9.72s on master (3% faster)
```

**Comment by ornithocoder** [comment]
@marcelofabri I'll try to cache the result. Changing it to WIP for now.

**Comment by ornithocoder** [comment]
@marcelofabri I've tried cache, but got the same results. Cache implementation can be found in this gist https://gist.github.com/ornithocoder/eb4b407fdbd8e74cbeb3d7bdb0254bd9.

The result is pretty much the same, but honestly, I prefer the cache implementation better and am thinking of pushing it. What do you think?

**Comment by marcelofabri** [comment]
Since the results are the same, I wouldn't add the cache to keep things simpler. One thing that *might* help is to only evaluate `let isEmpty = line.content.trimmingCharacters(in: .whitespaces).isEmpty` if `comment` is `true`.

**Comment by ornithocoder** [comment]
@marcelofabri I decided to give it another try and increase the number of iterations to 20. These are the results:

**Without Caching**

```
personal-fork-swiftlint $ ./script/oss-check --branch blank_line --iteration 20
Cloning Aerial
Cloning Alamofire
Cloning Firefox
Cloning Kickstarter
Cloning Moya
Cloning Nimble
Cloning Quick
Cloning Realm
Cloning SourceKitten
Cloning Sourcery
Cloning Swift
Cloning WordPress
Building branch
Preparing osscheck/builds (identifier builds)
Linting 20 iterations of Aerial with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Alamofire with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Firefox with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Kickstarter with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Moya with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Nimble with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Quick with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Realm with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of SourceKitten with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Sourcery with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Swift with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of WordPress with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Building master
Previous HEAD position was 9959280a... Add .testFunctionBodyLengthsWithIfs() to Linux tests
HEAD is now at ef67adca... rename RulesDocsOptions to GenerateDocsOptions
Linting 20 iterations of Aerial with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Alamofire with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Firefox with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Kickstarter with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Moya with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Nimble with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Quick with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Realm with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of SourceKitten with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Sourcery with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Swift with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of WordPress with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Message: Linting Aerial with this PR took 0.31s vs 0.31s on master (0% slower)
Message: Linting Alamofire with this PR took 2.43s vs 2.39s on master (1% slower)
Message: Linting Firefox with this PR took 10.49s vs 10.12s on master (3% slower)
Message: Linting Kickstarter with this PR took 15.24s vs 15.44s on master (1% faster)
Message: Linting Moya with this PR took 1.02s vs 0.7s on master (45% slower)
Message: Linting Nimble with this PR took 1.4s vs 1.32s on master (6% slower)
Message: Linting Quick with this PR took 0.42s vs 0.44s on master (4% faster)
Message: Linting Realm with this PR took 2.23s vs 2.38s on master (6% faster)
Message: Linting SourceKitten with this PR took 0.8s vs 0.78s on master (2% slower)
Message: Linting Sourcery with this PR took 3.94s vs 3.86s on master (2% slower)
Message: Linting Swift with this PR took 11.08s vs 11.01s on master (0% slower)
Message: Linting WordPress with this PR took 10.0s vs 10.04s on master (0% faster)
```

**With Caching**

```
$ ./script/oss-check --branch blank_line --iteration 20
Cloning Aerial
Cloning Alamofire
Cloning Firefox
Cloning Kickstarter
Cloning Moya
Cloning Nimble
Cloning Quick
Cloning Realm
Cloning SourceKitten
Cloning Sourcery
Cloning Swift
Cloning WordPress
Building branch
Preparing osscheck/builds (identifier builds)
Linting 20 iterations of Aerial with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Alamofire with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Firefox with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Kickstarter with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Moya with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Nimble with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Quick with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Realm with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of SourceKitten with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Sourcery with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Swift with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of WordPress with branch: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Building master
Previous HEAD position was 369c4f18... Cache again
HEAD is now at ef67adca... rename RulesDocsOptions to GenerateDocsOptions
Linting 20 iterations of Aerial with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Alamofire with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Firefox with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Kickstarter with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Moya with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Nimble with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Quick with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Realm with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of SourceKitten with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Sourcery with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of Swift with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Linting 20 iterations of WordPress with master: 1..2..3..4..5..6..7..8..9..10..11..12..13..14..15..16..17..18..19..20
Message: Linting Aerial with this PR took 0.33s vs 0.32s on master (3% slower)
Message: Linting Alamofire with this PR took 1.44s vs 1.57s on master (8% faster)
Message: Linting Firefox with this PR took 10.37s vs 11.21s on master (7% faster)
Message: Linting Kickstarter with this PR took 15.48s vs 16.28s on master (4% faster)
Message: Linting Moya with this PR took 0.99s vs 0.83s on master (19% slower)
Message: Linting Nimble with this PR took 1.29s vs 1.45s on master (11% faster)
Message: Linting Quick with this PR took 0.49s vs 0.46s on master (6% slower)
Message: Linting Realm with this PR took 2.19s vs 2.36s on master (7% faster)
Message: Linting SourceKitten with this PR took 0.75s vs 0.91s on master (17% faster)
Message: Linting Sourcery with this PR took 3.83s vs 3.94s on master (2% faster)
Message: Linting Swift with this PR took 11.03s vs 11.29s on master (2% faster)
Message: Linting WordPress with this PR took 9.95s vs 9.55s on master (4% slower)
```

**Comment by ornithocoder** [comment]
@marcelofabri @jpsim this one is ready for code review. I changed #1747 to WIP because I want to rebase that one once the number of lines issue is fixed.

**Comment by marcelofabri** [comment]
@ornithocoder Just to let you know that I didn't forget about this one. However, this needs a more careful review and I haven't got time yet. 😬

**Comment by ornithocoder** [comment]
This PR had several commits, so I merged some commits leaving only two:

* 403e9632d7e6ed34518ce7687cc2d894e6726549 - Fix number of lines reported by rules
  * Updates SwiftLint code to fix the number of lines issue.
* 0fbef60189fd0df70cbe905defcc0eef9bf0eb6b - Reduce number of lines of code
  * Minor changes on SwiftLint's code to respect the number of lines of code specified for the project.

**Comment by ornithocoder** [comment]
@marcelofabri I've rebased so it shows the performance impact before all the other messages.

**Comment by ornithocoder** [comment]
Hi @marcelofabri @jpsim, any updates on this one? I know it looks a long PR, but in fact it isn't. The first commit contains the real changes to improve the way the number of lines is calculated. The second commit is just to adapt SwiftLint's codebase so function_body_length and type_body_length don't throw any violations.

If there's anything I could do to make it easier to review, please let me know.

**Comment by ornithocoder** [comment]
@marcelofabri @jpsim should I keep rebasing to resolve the CHANGELOG.md conflict or wait for the review first?

**Comment by marcelofabri** [comment]
I thought about it and I think we probably should raise the default limits instead of applying those fixes. Most of them don't make code easier to read IMO. 

Or we can even keep the current behavior of considering things that are not tokens (`}`) not relevant and just improve the messages.

I'd love to hear @jpsim's thoughts on this too.

**Comment by ornithocoder** [comment]
Then I would recommend removing

* function_body_length
* type_body_length

Right now the number of lines they report doesn't reflect the valid number of lines of code.

And not applying this fix - or a different solution for the same issue - also makes #52 impossible.

**Comment by ornithocoder** [comment]
I like the idea of applying the first commit (fix) but increasing the default values ;-)

**Comment by ornithocoder** [comment]
@marcelofabri your suggestion would be to increase the defaults values for the function_body_length and type_body_length rules or to bump the number of lines on SwiftLint's .swiftlint.yml?

**Comment by ornithocoder** [comment]
I've update the PR w/ fix + new default values. I still have the old commit in a different branch, just in case ;-) @marcelofabri @jpsim

**Comment by ornithocoder** [comment]
@jpsim what's your opinion on increasing the default values for function_body_length from 40 to 50 and type_body_length from 200 to 250?

**Comment by ornithocoder** [comment]
Hi folks. Any change of getting this into master?

**Comment by ornithocoder** [comment]
@jpsim @marcelofabri `function_body_length` and `type_body_length` are currently reporting the wrong number of lines. Any chance of getting this into master? It's a simple change with pretty much no performance impact. For the new default values I've used SwiftLint as baseline. Oh, and it also unblocks me to submit a pull request for #52. Thanks.

**Comment by jpsim** [comment]
Finally took a first pass at reviewing this. I agree that if caching results doesn't impact performance, we simply shouldn't cache results to simplify this change.

Also, this seems overly complex for what it's trying to do, but maybe I'm missing something?

Why isn't this PR as simple as just this:

```swift
fileprivate func numberOfCommentAndWhitespaceOnlyLines(startLine: Int, endLine: Int) -> Int {
    let commentKinds = SyntaxKind.commentKinds
    let linesAndSyntaxKinds = zip(lines[(startLine - 1)...(endLine - 1)], syntaxKindsByLines[startLine...endLine])
    return linesAndSyntaxKinds.filter { line, kinds in
        let whitespaceOnly = line.content.trimmingCharacters(in: .whitespaces).isEmpty
        if whitespaceOnly { return true }
        let commentOnly = kinds.filter { !commentKinds.contains($0) }.isEmpty
        return commentOnly
    }.count
}
```

**Comment by ornithocoder** [comment]
@jpsim thanks! That's pretty close to the original implementation we had, but for some stupid reason I decided to insist on the cached version :-) I'm pushing a new version that doesn't cache.

File: Source/SwiftLintFramework/Rules/FunctionBodyLengthRule.swift [hunk 1]
```diff
@@ -9,7 +9,7 @@
 import SourceKittenFramework
 
 public struct FunctionBodyLengthRule: ASTRule, ConfigurationProviderRule {
-    public var configuration = SeverityLevelsConfiguration(warning: 40, error: 100)
+    public var configuration = SeverityLevelsConfiguration(warning: 50, error: 100)
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/FunctionBodyLengthRule.swift]: What's the motivation for this change and why isn't it noted in the changelog?

File: Source/SwiftLintFramework/Rules/FileHeaderRule.swift [hunk 2]
```diff
@@ -51,9 +51,7 @@ public struct FileHeaderRule: ConfigurationProviderRule, OptInRule {
             }
 
             // skip SwiftLint commands
-            guard !isSwiftLintCommand(token: token, file: file) else {
-                continue
-            }
+            guard !isSwiftLintCommand(token: token, file: file) else { continue }
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/FileHeaderRule.swift]: This doesn't help with readability.

File: Source/SwiftLintFramework/Extensions/File+SwiftLint.swift [hunk 3]
```diff
@@ -270,8 +270,11 @@ extension File {
 
     fileprivate func numberOfCommentAndWhitespaceOnlyLines(startLine: Int, endLine: Int) -> Int {
         let commentKinds = SyntaxKind.commentKinds
-        return syntaxKindsByLines[startLine...endLine].filter { kinds in
-            kinds.filter { !commentKinds.contains($0) }.isEmpty
+        let linesAndSyntaxKinds = zip(lines[(startLine - 1)...(endLine - 1)], syntaxKindsByLines[startLine...endLine])
+
+        return linesAndSyntaxKinds.filter { line, kinds in
+            let isLineEmpty = line.content.trimmingCharacters(in: .whitespaces).isEmpty
+            return isLineEmpty ? true : !kinds.filter { commentKinds.contains($0) }.isEmpty
```
Comment (jpsim) [inline Source/SwiftLintFramework/Extensions/File+SwiftLint.swift]: As a general rule, `filter { contains }.isEmpty` can always be restructured to `contains { contains }`

File: Source/SwiftLintFramework/Rules/FunctionBodyLengthRule.swift (hunk 1 above)
Comment (ornithocoder) [reply Source/SwiftLintFramework/Rules/FunctionBodyLengthRule.swift]: The number of violations it throws for the SwiftLint codebase. [See here](https://github.com/realm/SwiftLint/pull/1753#issuecomment-329599707).

File: Source/SwiftLintFramework/Rules/FileHeaderRule.swift (hunk 2 above)
Comment (ornithocoder) [reply Source/SwiftLintFramework/Rules/FileHeaderRule.swift]: I agree. Any suggestion to decrease the number of lines there to avoid throwing a violation because of the number of lines?

File: Source/SwiftLintFramework/Extensions/File+SwiftLint.swift (hunk 3 above)
Comment (ornithocoder) [reply Source/SwiftLintFramework/Extensions/File+SwiftLint.swift]: I'll try to improve it.

File: Source/SwiftLintFramework/Rules/FunctionBodyLengthRule.swift (hunk 1 above)
Comment (jpsim) [reply Source/SwiftLintFramework/Rules/FunctionBodyLengthRule.swift]: You can override default configuration values in `.swiftlint.yml`.

File: Source/SwiftLintFramework/Rules/FileHeaderRule.swift (hunk 2 above)
Comment (jpsim) [reply Source/SwiftLintFramework/Rules/FileHeaderRule.swift]: Could split up into multiple functions.

File: Source/SwiftLintFramework/Extensions/File+SwiftLint.swift (hunk 3 above)
Comment (ornithocoder) [reply Source/SwiftLintFramework/Extensions/File+SwiftLint.swift]: Should be fixed now. Thanks!

File: Source/SwiftLintFramework/Rules/FunctionBodyLengthRule.swift (hunk 1 above)
Comment (ornithocoder) [reply Source/SwiftLintFramework/Rules/FunctionBodyLengthRule.swift]: Done

File: Source/SwiftLintFramework/Rules/FileHeaderRule.swift (hunk 2 above)
Comment (ornithocoder) [reply Source/SwiftLintFramework/Rules/FileHeaderRule.swift]: Reverted.

File: Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift [hunk 4]
```diff
@@ -21,15 +21,15 @@ private func violatingFuncWithBody(_ body: String) -> String {
...
+        let longFunctionBody = funcWithBody(repeatElement("x = 0\n", count: 38).joined())
         XCTAssertEqual(violations(longFunctionBody), [])
 
         let longerFunctionBody = violatingFuncWithBody(repeatElement("x = 0\n", count: 40).joined())
         XCTAssertEqual(violations(longerFunctionBody), [StyleViolation(
             ruleDescription: FunctionBodyLengthRule.description,
             location: Location(file: nil, line: 1, character: 1),
             reason: "Function body should span 40 lines or less excluding comments and " +
-            "whitespace: currently spans 41 lines")])
+            "whitespace: currently spans 42 lines")])
```
Comment (jpsim) [inline Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift]: Sorry, I fail to see how this was previously incorrect and is now correct.

We're adding 40 lines to the function body additional to its first line:

```swift
violatingFuncWithBody(repeatElement("x = 0\n", count: 40).joined())
```

So the body should be considered to span 41 lines.

Where are you getting 42 from?

File: Source/SwiftLintFramework/Extensions/File+SwiftLint.swift [hunk 5]
```diff
@@ -270,8 +270,11 @@ extension File {
 
     fileprivate func numberOfCommentAndWhitespaceOnlyLines(startLine: Int, endLine: Int) -> Int {
         let commentKinds = SyntaxKind.commentKinds
-        return syntaxKindsByLines[startLine...endLine].filter { kinds in
-            kinds.filter { !commentKinds.contains($0) }.isEmpty
+        let linesAndSyntaxKinds = zip(lines[(startLine - 1)...(endLine - 1)], syntaxKindsByLines[startLine...endLine])
```
Comment (jpsim) [inline Source/SwiftLintFramework/Extensions/File+SwiftLint.swift]: How about this?

```swift
return (startLine...endLine).filter { lineNumber in
    let line = lines[lineNumber - 1]
    let isLineEmpty = line.content.rangeOfCharacter(from: CharacterSet.whitespaces.inverted) == nil
    if isLineEmpty { return true }

    return syntaxKindsByLines[lineNumber].contains(where: SyntaxKind.commentKinds.contains)
}.count
```

It's simpler (IMHO) and more optimized because it avoids unconditionally iterating over all characters in the string, allocating a new string, and mutating it once for every whitespace character.

File: Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift (hunk 4 above)
Comment (ornithocoder) [reply Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift]: screen shot 2017-11-14 at 9 49 56 am

I didn't touch the logic in 

```swift
    internal func exceedsLineCountExcludingCommentsAndWhitespace(_ start: Int, _ end: Int,
                                                                 _ limit: Int) -> (Bool, Int) {
        guard end - start > limit else {
            return (false, end - start)
        }

        let count = end - start - numberOfCommentAndWhitespaceOnlyLines(startLine: start, endLine: end)
        return (count > limit, count)
    }
```

with takes the endLine - startLine.

File: Source/SwiftLintFramework/Extensions/File+SwiftLint.swift (hunk 5 above)
Comment (ornithocoder) [reply Source/SwiftLintFramework/Extensions/File+SwiftLint.swift]: Done

File: Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift (hunk 4 above)
Comment (jpsim) [reply Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift]: Except that the counts _did_ change, you even had to change the tests to accommodate it.

If you look at your screenshot, the body spans from line 2 to line 42, meaning 41 lines.

Comment (ornithocoder) [reply Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift]: That's because the last line contains a single "}" and was computed as empty line :-)

Comment (ornithocoder) [reply Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift]: So the problem now isn't the `numberOfCommentAndWhitespaceOnlyLines(startLine:endLine)`. it now reports the correct number of lines that don't have code. Rules that use `exceedsLineCountExcludingCommentsAndWhitespace(_:_:_:)` have to be changed to pass the correct start line and end line of blocks.

Comment (jpsim) [reply Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift]: I wouldn't count that as the body. i.e. this function should have a body line count of 2:

```swift
func foo() { // not body
  _ = 0 // body line 1
  
  // not body, neither is above
  _ = 0 // body line 2
} // not body
```

Comment (ornithocoder) [reply Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift]: I agree, but not it's a different problem.

* `numberOfCommentAndWhitespaceOnlyLines`, and 
* `exceedsLineCountExcludingCommentsAndWhitespace`

report the correct number of lines. What needs to be changed now are the rules that use `exceedsLineCountExcludingCommentsAndWhitespace`.

The only reason why the rule was reporting the correct number of lines (41) was because `numberOfCommentAndWhitespaceOnlyLines` was reporting the wrong of valid lines of code. One bug was hiding the other.

Next step is to review all `_length` rules to pass the correct startLine and endLine to `exceedsLineCountExcludingCommentsAndWhitespace`exceedsLineCountExcludingCommentsAndWhitespace.

Comment (ornithocoder) [reply Tests/SwiftLintFrameworkTests/FunctionBodyLengthRuleTests.swift]: I'll squash some of these the commits as "Fixes numberOfCommentAndWhitespaceOnlyLines" and will changed it to WIP, while the rules are not fixed.

File: Source/SwiftLintFramework/Extensions/File+SwiftLint.swift [hunk 6]
```diff
@@ -281,7 +284,7 @@ extension File {
             return (false, end - start)
         }
 
-        let count = end - start - numberOfCommentAndWhitespaceOnlyLines(startLine: start, endLine: end)
+        let count = end - start - numberOfCommentAndWhitespaceOnlyLines(startLine: start, endLine: end) - 1
```
Comment (ornithocoder) [inline Source/SwiftLintFramework/Extensions/File+SwiftLint.swift]: Next step is to make

`let count = (start...end).count - numberOfCommentAndWhitespaceOnlyLines(startLine: start, endLine: end)` to include startLine and endLine and update the `function_body_length` and `type_body_length` to pass new startLine and endLine values.

**Comment by ornithocoder** [comment]
Hi @jpsim @marcelofabri, since the method to count the number of lines doesn't understand context (and it shouldn't, so it can be reused for type length, closure length, function length, etc...), it's important to call it with the correct startLine and finalLine.

I'm struggling to find the correct startLine and finalLine for the rule `function_body_length`. Right now it passes the line where the function starts (line containing the "{") and line where the function ends "}". But by passing these two lines, they are counted are valid lines. For instance,

passing 1 and 3 for the code below

```
1 func foo() {
2
3 }
```

it returns two valid lines - not empty and not comment - (1 and 3) and one line with empty content (2).

So I'm looking for a way to changed the `function_body_length` rule to pass the actual line where the function content starts. For instance:

```
1 func foo() {
2
3 }

firstLine = 2, lastLine = 2
total number of lines = (2...2).count = 1, valid lines = 0
```

```
1 func foo() {
2   _ = 0
3
4   _ = 0
5 }

firstLine = 2, lastLine = 4
total number of lines = (2...4).count = 3, valid lines = 2
```

```
1 func foo() {
2   _ = 0
3
4   _ = 0 }

firstLine = 2, lastLine = 4
total number of lines = (2...4).count = 3, valid lines = 2
```

```
1 func foo() { }

firstLine = 1, lastLine = 1
total number of lines = (1...1).count = 1
```

```
1 func foo() -> Int { return 0 }

firstLine = 1, lastLine = 1
total number of lines = (1...1).count = 1
```

Any idea how can I achieve this?

Recap: 

1. `numberOfCommentAndWhitespaceOnlyLines(startLine:endLine)` was not counting lines with "{" and "}", so for long functions with ifs, elses, guards, etc... the number of valid lines was way off.
2. Fixing `numberOfCommentAndWhitespaceOnlyLines(startLine:endLine)` highlighted another issue, that the rules are not passing the correct start line and end line. Since { and } where not computed by the first issue, this was never a problem. But with the fix, the last line (which contains a single "}") is computed causing all the issues @jpsim noticies above on the code review.
3. I need help to fix the `function_body_length` to pass the correct startLine and endLine.

I've pushed a workaround that includes a `inclusive` flag so functions using the `numberOfCommentAndWhitespaceOnlyLines` can specify if they want to count the first and last line or not. But this is a workaround. I would like to pass the correct startLine and finalLine instead.

Thanks.

**Comment by ornithocoder** [comment]
I'll have to find another way to

1) fix the number of lines reported
2) find where a function starts and ends.

I'm abandoning this PR to start fresh on this issue.
